### PR TITLE
Update: Implement FlatConfigArray (refs #13481)

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

Implements FlatConfigArray (refs #13481)

#### What changes did you make? (Give an overview)

I added `FlatConfigArray` class and associated tests to make sure everything is working okay. `FlatConfigArray` is not referenced anywhere while running ESLint currently, as implementing `FlatConfigArray` is meant to be a standalone task. In the next task, I'll work on using it when `eslint.config.js` is found.

Merging this does *not* result in any end-user-facing changes. It will result in `FlatConfigArray` tests being run with `npm test`

#### Is there anything you'd like reviewers to focus on?

1. Are there any cases I missed?
2. Is the code easy enough to follow? (The old config system was so complicated that the code was hard to follow, I want this version to be easier for people to look at and figure out what's going on.)

### CODE REVIEW

File: lib/config/flat-config-schema.js [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+
+
+/**
+ * Validates that a given string is the form pluginName/objectName.
+ * @param {string} value The string to check.
+ * @returns {void}
+ * @throws {TypeError} If the string isn't in the correct format.
+ */
+function assertIsPluginMemberName(value) {
+    if (!/[@a-z0-9-_$]+\/[a-z0-9-_$]+/iu.test(value)) {
```
Comment (btmills) [inline lib/config/flat-config-schema.js:77]: Are we intentionally precluding scoped package-style plugin names here? I realize that the plugin's name is an arbitrary object key that doesn't need to match the npm package name, but I wouldn't be surprised if matching package names becomes a common pattern among users for simplicity.

File: lib/config/flat-config-schema.js [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+            throw new TypeError("Expected \"script\", \"module\", or \"commonjs\".");
+        }
+    }
+};
+
+//-----------------------------------------------------------------------------
+// Full schema
+//-----------------------------------------------------------------------------
+
+exports.flatConfigSchema = {
```
Comment (btmills) [inline lib/config/flat-config-schema.js:352]: This PR doesn't include the `name` key that was included in the RFC. Is that an intentional omission?

File: lib/config/default-config.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    has(target, property) {
+        return Rules.has(property);
+    }
+});
+
+exports.defaultConfig = [
+    {
+        plugins: {
+            "@": {
```
Comment (btmills) [inline lib/config/default-config.js:36]: Is there any reason we'd want this to be a `Symbol` rather than a special string?

File: lib/config/flat-config-schema.js [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+const globalsSchema = {
+    merge: "assign",
+    validate(value) {
+
+        assertIsObject(value);
+
+        for (const key of Object.keys(value)) {
+
+            // avoid hairy edge case
+            if (key === "__proto__") {
```
Comment (btmills) [inline lib/config/flat-config-schema.js:142]: This check occurs in a number of places. Is this class of edge cases sufficiently hairy that it's worth looking into an implementation (like using `Map`s internally instead) that avoids them entirely? Alternatively, are we comfortable with a few footguns lying around in this file since we ideally won't need to visit it often once this is merged?

**Review by btmills — COMMENTED** [review]
I am pleasantly surprised by how much less is going on here than in the existing config system. Outsourcing all of the file path matching to `@humanwhocodes/config-array` nicely separates the "what needs to be merged" logic from the "what are the rules for merging each field"  logic. It's great to see the vision of a simpler config coming together :tada:

File: lib/config/flat-config-schema.js (hunk 2 above)
Comment (nzakas) [reply lib/config/flat-config-schema.js:352]: Yes. The `name` key is one of the default keys that are automatically included with `ConfigArray`. The other two are `files` and `ignores`. These three are used internally so they need a very specific validation/merge strategy for everything to work, hence, they are defined by default.

File: lib/config/flat-config-schema.js (hunk 4 above)
Comment (nzakas) [reply lib/config/flat-config-schema.js:142]: I copied this check over from the eslintrc functionality figuring it was probably added for a reason. I think switching to maps at this point is probably premature and will add complexity that isn't really necessary for the number of times this edge case might occur.

File: lib/config/flat-config-schema.js (hunk 1 above)
Comment (nzakas) [reply lib/config/flat-config-schema.js:77]: I suppose I was assuming that people would stop using the fully scoped package name in the plugin reference once the config was in JavaScript. It probably doesn't hurt to allow it, though.

File: tests/lib/config/flat-config-array.js [hunk 5]
```diff
@@ -0,0 +1,1153 @@
...
+
+                await assertInvalidConfig([
+                    {
+                        languageOptions: {
+                            foo: true
+                        }
+                    }
+                ], "Unexpected key \"foo\" found.");
+
+            });
```
[the remark this answers is not in the document]
Comment (nzakas) [reply tests/lib/config/flat-config-array.js:524]: I'll add it. I was struggling to come up with all possible variations.

**Comment by nzakas** [comment]
I believe everything has been addressed, but let me know if I missed anything.

File: lib/config/flat-config-schema.js (hunk 2 above)
Comment (btmills) [reply lib/config/flat-config-schema.js:352]: Ah, I missed that `name` is a first-class part of `ConfigArray`. Thanks for explaining.

File: lib/config/flat-config-schema.js (hunk 4 above)
Comment (btmills) [reply lib/config/flat-config-schema.js:142]: I couldn’t find any occurrence of these checks prior to https://github.com/eslint/eslint/pull/11546, so I’ll assume they were introduced there to guard against prototype pollution. I have no problem leaving this as-is given it’s been this way for a couple years and is only rarely touched.

File: lib/config/flat-config-schema.js [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+
+/** @type {ObjectPropertySchema} */
+const processorSchema = {
+    merge: "replace",
+    validate(value) {
+        if (typeof value === "string") {
+            assertIsPluginMemberName(value);
+        } else if (value && typeof value === "object") {
+            if (typeof value.preprocess !== "function" || typeof value.postprocess !== "function") {
+                throw new TypeError("Object must have a preprocess() and a postprocess() method.");
```
Comment (btmills) [inline lib/config/flat-config-schema.js:226]: There's not currently a test exercising this check.

File: lib/config/flat-config-schema.js [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+        // manually validate that plugins are not redefined
+        for (const key of keys) {
+
+            // avoid hairy edge case
+            if (key === "__proto__") {
+                continue;
+            }
+
+            if (key in first && key in second && first[key] !== second[key]) {
+                throw new TypeError(`Cannot redefine plugin "${key}".`);
```
Comment (btmills) [inline lib/config/flat-config-schema.js:188]: There's not currently a test exercising this check.

File: lib/config/flat-config-schema.js [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+
+        for (const key of Object.keys(value)) {
+
+            // avoid hairy edge case
+            if (key === "__proto__") {
+                continue;
+            }
+
+            if (key !== key.trim()) {
+                throw new TypeError(`Global "${key}" has leading or trailing whitespace.`);
```
Comment (btmills) [inline lib/config/flat-config-schema.js:147]: There's not currently a test exercising this check.

File: lib/config/flat-config-array.js [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+            languageOptions.parser = plugins[pluginName].parsers[parserName];
+        }
+
+        // Check processor value
+        if (processor && typeof processor === "string") {
+            const { pluginName, objectName: processorName } = splitPluginIdentifier(processor);
+
+            if (!plugins || !plugins[pluginName] || !plugins[pluginName].processors || !plugins[pluginName].processors[processorName]) {
+                throw new TypeError(`Key "processor": Could not find "${processorName}" in plugin "${pluginName}".`);
```
Comment (btmills) [inline lib/config/flat-config-array.js:111]: There's not currently a test exercising this check.

**Review by btmills — APPROVED** [review]
I just did a thorough pass through the schema to make sure everything matches up. I'm excited that we'll be able to alert users to invalid configurations as close to the source as possible with some of the checks you've added. There were a few particularly helpful ones that aren't covered by a test, so I pointed those out in case you'd like to add those. LGTM either way.

**Comment by nzakas** [comment]
Nice! I’ll plan on adding tests for those.

File: lib/config/flat-config-schema.js (hunk 6 above)
Comment (nzakas) [reply lib/config/flat-config-schema.js:226]: Looks like it's tested on line 369. (For some reason, can't get a direct link.)

File: lib/config/flat-config-schema.js (hunk 7 above)
Comment (nzakas) [reply lib/config/flat-config-schema.js:188]: Tested on line 260.

File: lib/config/flat-config-schema.js (hunk 6 above)
Comment (btmills) [reply lib/config/flat-config-schema.js:226]: I was searching for the whole string, but I see the test covers the important part. It's quite possible I made the same mistake for the other checks I commented on.

File: lib/config/flat-config-schema.js (hunk 8 above)
Comment (nzakas) [reply lib/config/flat-config-schema.js:147]: Added

File: lib/config/flat-config-array.js (hunk 9 above)
Comment (nzakas) [reply lib/config/flat-config-array.js:111]: Added

**Review by btmills — APPROVED** [review]
Nice! Sorry about the false positive on a couple of those.

File: lib/config/flat-config-array.js [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+ */
+class FlatConfigArray extends ConfigArray {
+
+    /**
+     * Creates a new instance.
+     * @param {*[]} configs An array of configuration information.
+     * @param {{basePath: string, baseConfig: FlatConfig}} options The options
+     *      to use for the config array instance.
+     */
+    constructor(configs, { basePath, baseConfig = defaultConfig }) {
```
Comment (mdjermanovic) [inline lib/config/flat-config-array.js:55]: Is `baseConfig` parameter necessary, shouldn't we always use `defaultConfig` as the first element?

File: lib/config/default-config.js [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+                parsers: {
+                    espree: require("espree")
+                },
+                rules: defaultRulesConfig
+            }
+        },
+        ignores: ["node_modules/**"],
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/config/default-config.js:43]: These are all glob patterns now. I’ll update them to match the current defaults.

File: lib/config/flat-config-array.js (hunk 10 above)
Comment (nzakas) [reply lib/config/flat-config-array.js:55]: I use this in the tests to avoid needing to load the entire default config when I need to test specific functionality that shouldn’t be impacted by any changes to the default config.

File: lib/config/default-config.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (mdjermanovic) [reply lib/config/default-config.js:43]: Does it also mean that we'll interpret the content of `.eslintignore` file differently, depending on the format of the found eslint configuration file: 

* as a list of `.gitignore` patterns (actual behavior) if the config is `.eslintrc.*`.
* as a list of glob patterns if the config is `eslint.config.js`

File: lib/config/flat-config-array.js [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+ * @author Nicholas C. Zakas
+ */
+
+"use strict";
+
+//-----------------------------------------------------------------------------
+// Requirements
+//-----------------------------------------------------------------------------
+
+const { ConfigArray, ConfigArraySymbol } = require("@humanwhocodes/config-array");
```
Comment (mdjermanovic) [inline lib/config/flat-config-array.js:12]: Will `ConfigArray` support negated patterns (prefixed with `!`) in `files` and/or `ignores`?

File: lib/config/default-config.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/config/default-config.js:43]: No. There is no change to how .eslintignore will be interpreted. It will be read and compiled into a function that is then included in a flat config ignores key. (Can be either a glob pattern or a function.)

File: lib/config/flat-config-array.js (hunk 12 above)
Comment (nzakas) [reply lib/config/flat-config-array.js:12]: It already does

Comment (mdjermanovic) [reply lib/config/flat-config-array.js:12]: What's the expected behavior in this case:

```js
{
    files: ["*.js", "!foo.js"]
}
```

By the RFC, it might be expected that the config applies to all `*.js` files except `foo.js`:

> These patterns can be negated by prefixing them with `!`, which effectively mimics the behavior of `excludedFiles` in `.eslintrc`.

However, `ConfigArray` checks if the given file path matches at least one of the patterns in the `files` array, so the config will apply to any file path (`a.js`, `foo.js`, `a.ts`, `package.json`...).

File: lib/config/flat-config-schema.js [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+
+
+/**
+ * Validates that a given string is the form pluginName/objectName.
+ * @param {string} value The string to check.
+ * @returns {void}
+ * @throws {TypeError} If the string isn't in the correct format.
+ */
+function assertIsPluginMemberName(value) {
+    if (!/[@a-z0-9-_$]+\/(?:[a-z0-9-_$]+)+/iu.test(value)) {
```
Comment (mdjermanovic) [inline lib/config/flat-config-schema.js:77]: The second `+` in `(?:[a-z0-9-_$]+)+` looks unnecessary (and might cause catastrophic backtracking if we add `$` anchor or anything else after `+` in the pattern), did you mean something else?

File: lib/config/flat-config-schema.js [hunk 14]
```diff
@@ -0,0 +1,362 @@
...
+    [0, 0], ["off", 0],
+    [1, 1], ["warn", 1],
+    [2, 2], ["error", 2]
+]);
+
+const globalVariablesValues = new Set([
+    true, "writable", "writeable",
+    false, "readonly",
+    "off"
+]);
```
Comment (mdjermanovic) [inline lib/config/flat-config-schema.js:34]: Aside from these, we also allow: `"true"` (string), `null`, `"false"` (string), and `"readable"`.

https://github.com/eslint/eslintrc/blob/62ea4bd74b78fbeff12ffb21f1f978817601d4d1/lib/shared/config-ops.js#L108-L129

File: lib/config/flat-config-schema.js [hunk 15]
```diff
@@ -0,0 +1,362 @@
...
+            if (key === "__proto__") {
+                continue;
+            }
+
+            if (key !== key.trim()) {
+                throw new TypeError(`Global "${key}" has leading or trailing whitespace.`);
+            }
+
+            if (!globalVariablesValues.has(value[key])) {
+                throw new TypeError("Expected \"readonly\", \"writeable\", or \"off\".");
```
Comment (mdjermanovic) [inline lib/config/flat-config-schema.js:151]:
```suggestion
                throw new TypeError("Expected \"readonly\", \"writable\", or \"off\".");
```

File: lib/config/flat-config-array.js (hunk 12 above)
Comment (nzakas) [reply lib/config/flat-config-array.js:12]: This behavior is pretty well-defined in the RFC: `files` are treated as OR statements. If you want to do an AND, then you need to group them in an array:

```
{
    files: [["*.js", "!foo.js"]]
}
```

File: lib/config/flat-config-schema.js (hunk 15 above)
Comment (nzakas) [reply lib/config/flat-config-schema.js:151]: "writable" is a misspelling, so we don't document that even though we support it. (This option was incorrectly implemented with the misspelling initially.)

File: lib/config/flat-config-schema.js (hunk 13 above)
Comment (nzakas) [reply lib/config/flat-config-schema.js:77]: Ooh nice catch. Yeah, the intent was to allow `foo/bar/baz` to map what we have in eslintrc, but clearly I messed that pattern up.

**Comment by nzakas** [comment]
Comments addressed.

**Comment by mdjermanovic** [comment]
@nzakas regarding https://github.com/eslint/eslint/pull/14321#discussion_r622205891, isn't it the opposite (`writable` - correct, `writeable` - misspelling)?

Our [documentation](https://eslint.org/docs/user-guide/configuring/language-options#specifying-globals) uses `"writable"`, and says that `"writeable"` can be used for historical reasons.

File: lib/config/default-config.js [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+            "@": {
+                parsers: {
+                    espree: require("espree")
+                },
+                rules: defaultRulesConfig
+            }
+        },
+        ignores: [
+            "**/node_modules/**",
+            ".*"
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/config/default-config.js:45]: I copied this from the eslintrc default ignore patterns for consistency.

File: lib/config/flat-config-schema.js [hunk 17]
```diff
@@ -0,0 +1,362 @@
...
+
+const ruleSeverities = new Map([
+    [0, 0], ["off", 0],
+    [1, 1], ["warn", 1],
+    [2, 2], ["error", 2]
+]);
+
+const globalVariablesValues = new Set([
+    true, "true", "writable", "writeable",
+    false, "false", "readonly", "readable",
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/config/flat-config-schema.js:32]: Geez, I had no idea that was allowed. But yes, we should support that for compat sake

File: lib/config/rule-validator.js [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+                const schema = getRuleOptionsSchema(rule);
+
+                if (schema) {
+                    this.validators.set(rule, ajv.compile(schema));
+                }
+            }
+
+            const validateRule = this.validators.get(rule);
+
+            if (validateRule && Array.isArray(config.rules[ruleId])) {
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/config/rule-validator.js:141]: Good point. I’ll update that.

File: lib/config/default-config.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (mdjermanovic) [reply lib/config/default-config.js:45]: Yes, but doesn't the RFC include the change in default patterns?

> With `eslint.config.js`, there are three ways that files and directories can be ignored:

> 1. **Defaults** - by default, ESLint will ignore `node_modules` and `.git` directories only. This is different from the current behavior where ESLint ignores `node_modules` and all files/directories beginning with a dot (`.`).

File: lib/config/rule-validator.js [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+     * @returns {void}
+     * @throws {Error} If a rule's configuration does not match its schema.
+     */
+    validate(config) {
+
+        if (!config.rules) {
+            return;
+        }
+
+        for (const ruleId of Object.keys(config.rules)) {
```
Comment (mdjermanovic) [inline lib/config/rule-validator.js:121]: We are currently not doing any validations for rules that are set to `0`/`"off"` severity.

This configuration is valid with the actual config system:

```js
rules: {
        "dot-location": [0, "foo"], // invalid option
        "foo": 0, // rule doesn't exist
        "foo/foo": 0, // plugin doesn't exist
    }
```

File: lib/config/default-config.js [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+//-----------------------------------------------------------------------------
+// Helpers
+//-----------------------------------------------------------------------------
+
+/*
+ * Because we try to delay loading rules until absolutely necessary, a proxy
+ * allows us to hook into the lazy-loading aspect of the rules map while still
+ * keeping all of the relevant configuration inside of the config array.
+ */
+const defaultRulesConfig = new Proxy({}, {
```
Comment (mdjermanovic) [inline lib/config/default-config.js:23]: Why we named this "Config"? It's an object with built-in rules, not a configuration?

File: lib/config/flat-config-array.js [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+        if (config === "eslint:recommended") {
+            return recommendedConfig;
+        }
+
+        if (config === "eslint:all") {
+            return allConfig;
+        }
+
+        return config;
+    }
```
Comment (mdjermanovic) [inline lib/config/flat-config-array.js:82]: If user makes a typo, like `"eslint:recomended"`, they'll get message `Error: All arguments must be objects.`

Might be nice to throw an error saying what `string` value was found and which `string` values are allowed?

File: lib/config/rule-validator.js [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+        pluginName = ruleIdParts.join("/");
+    }
+
+    if (!config.plugins || !config.plugins[pluginName]) {
+        throw new TypeError(`Key "rules": Key "${ruleId}": Could not find plugin "${pluginName}".`);
+    }
+
+    if (!config.plugins[pluginName].rules || !config.plugins[pluginName].rules[ruleName]) {
+        throw new TypeError(`Key "rules": Key "${ruleId}": Could not find "${ruleName}" in plugin "${pluginName}".`);
+    }
```
Comment (mdjermanovic) [inline lib/config/rule-validator.js:43]: Should we have a different message for built-in rules, this might be confusing:

`TypeError: Key "rules": Key "no-undeff": Could not find "no-undeff" in plugin "@".`

File: lib/config/default-config.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/config/default-config.js:45]: Huh. It appears so! Thanks for double-checking that. I seem to be caught in this maze of some areas where we wanted things to be consistent with eslintrc and some where they are different. Hard to keep track.

File: lib/config/flat-config-schema.js [hunk 23]
```diff
@@ -0,0 +1,362 @@
...
+            }
+
+            /*
+             * In any other situation, then the second rule config takes
+             * precedence. If it's an array, we return a copy;
+             * otherwise we return the value, which is just the severity.
+             */
+            result[ruleId] = secondIsArray
+                ? second[ruleId].slice(0)
+                : second[ruleId];
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/config/flat-config-schema.js:282]: Sorry, what is the "actual config system"?

File: lib/config/flat-config-schema.js [hunk 24]
```diff
@@ -0,0 +1,362 @@
...
+};
+
+/** @type {ObjectPropertySchema} */
+const processorSchema = {
+    merge: "replace",
+    validate(value) {
+        if (typeof value === "string") {
+            assertIsPluginMemberName(value);
+        } else if (value && typeof value === "object") {
+            if (typeof value.preprocess !== "function" || typeof value.postprocess !== "function") {
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/config/flat-config-schema.js:225]: eslintrc doesn't check for these, so you get a hairy error message if they are missing. I'd like to be more explicit in checking for these.

File: lib/config/flat-config-schema.js [hunk 25]
```diff
@@ -0,0 +1,362 @@
...
+            if (key === "__proto__") {
+                continue;
+            }
+
+            if (key !== key.trim()) {
+                throw new TypeError(`Global "${key}" has leading or trailing whitespace.`);
+            }
+
+            if (!globalVariablesValues.has(value[key])) {
+                throw new TypeError("Expected \"readonly\", \"writeable\", or \"off\".");
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/config/flat-config-schema.js:151]: Good idea :+1:

File: lib/config/flat-config-schema.js [hunk 26]
```diff
@@ -0,0 +1,362 @@
...
+            reportUnusedDisableDirectives: booleanSchema
+        }
+    },
+    languageOptions: {
+        schema: {
+            ecmaVersion: numberSchema,
+            sourceType: sourceTypeSchema,
+            globals: globalsSchema,
+            parser: parserSchema,
+            parserOptions: objectAssignSchema
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/config/flat-config-schema.js]: Fixed.

File: lib/config/rule-validator.js (hunk 22 above)
Comment (nzakas) [reply lib/config/rule-validator.js:43]: This is something we can address later. I don't want to tie up this PR one small UX discussions like this.

File: lib/config/default-config.js (hunk 20 above)
Comment (nzakas) [reply lib/config/default-config.js:23]: Fair point.

File: lib/config/flat-config-schema.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/config/flat-config-schema.js:349]: I'm guessing "actual config system" means eslintrc? Can you please use the term eslintrc instead, I've been having a hard time understanding these comments.

[the remark this answers is not in the document]
Comment (mdjermanovic) [reply lib/config/flat-config-schema.js:349]: Yes eslintrc, sorry for the confusion.

File: lib/config/flat-config-array.js (hunk 21 above)
Comment (nzakas) [reply lib/config/flat-config-array.js:82]: I'm going to hold off on addressing error cases until flat config is integrated into ESLint. We have a lot of options for messaging error conditions, and I don't think it's going to be clear what exactly is needed where until we can play with it.

File: lib/config/rule-validator.js (hunk 19 above)
Comment (nzakas) [reply lib/config/rule-validator.js]: Got it. Thanks.

File: lib/config/flat-config-schema.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/config/flat-config-schema.js]: Should be fixed now.

**Comment by nzakas** [comment]
I think I've fixed everything. @mdjermanovic thanks for the amazing level of detail in your comments.

File: tests/lib/config/flat-config-array.js [hunk 27]
```diff
@@ -0,0 +1,1413 @@
...
+ * @param {Object} rulesConfig The rules config portion of a config.
+ * @returns {Array} The rules config object.
+ */
+function normalizeRuleConfig(rulesConfig) {
+    for (const ruleId of Object.keys(rulesConfig)) {
+        rulesConfig[ruleId] = [2];
+    }
+
+    return rulesConfig;
+}
```
Comment (mdjermanovic) [inline tests/lib/config/flat-config-array.js:124]: Some tests unrelated to this PR [are failing](https://github.com/eslint/eslint/pull/14321/checks?check_run_id=2494731957), probably because this directly modifies the given object (`eslint-recommended` and `eslint-all`).

File: lib/config/flat-config-schema.js [hunk 28]
```diff
@@ -0,0 +1,444 @@
...
+
+    for (const key of Object.keys(result)) {
+
+        // avoid hairy edge case
+        if (key === "__proto__") {
+            continue;
+        }
+
+        const firstValue = first[key];
+        const secondValue = second[key];
```
Comment (mdjermanovic) [inline lib/config/flat-config-schema.js:81]: In some cases, this function will be called with `second === undefined`, so `second[key]` would crash.

Example:

```js
[
    {
        languageOptions: {
            parserOptions: {
                ecmaFeatures: {
                    jsx: true
                }
            }
        }
    },
    {
        languageOptions: {
            ecmaVersion: 2021
        }
    }
]
```

Throws `TypeError: Key "languageOptions": Key "parserOptions": Cannot read property 'jsx' of undefined`

File: lib/config/flat-config-schema.js [hunk 29]
```diff
@@ -0,0 +1,444 @@
...
+    return typeof value === "undefined";
+}
+
+/**
+ * Deeply merges two objects.
+ * @param {Object} first The base object.
+ * @param {Object} second The overrides object.
+ * @returns {Object} An object with properties from both first and second.
+ */
+function deepMerge(first, second) {
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/config/flat-config-schema.js:60]: Good catch. I think the expectation here is that an array overwrites an array. I'll double-check that.

File: tests/lib/config/flat-config-array.js (hunk 27 above)
Comment (nzakas) [reply tests/lib/config/flat-config-array.js:124]: Doh! Yup, I'll fix that. Thanks.

**Comment by nzakas** [comment]
Should be set now (hopefully).

File: lib/config/flat-config-schema.js [hunk 30]
```diff
@@ -0,0 +1,452 @@
...
+     * overwrite properties from the first. This sets up a baseline to use
+     * later rather than needing to inspect and change every property
+     * individually.
+     */
+    const result = {
+        ...first,
+        ...second
+    };
+
+    for (const key of Object.keys(second)) {
```
Comment (stephenwade) [inline lib/config/flat-config-schema.js:81]:
```suggestion
    for (const key of Object.keys(result)) {
```

If you only check keys in `second`, you fail to deeply clone objects in `first`. Since this returns a new object instead of modifying `first`, I think that all objects should be deeply cloned to prevent surprises.

```js
const o1 = { a: { x: 1, y: 2 } };
const o2 = { b: { z: 3 } };

const result = deepMerge(o1, o2);

result.b.z = 6;
console.log(o2.b.z) // 3 ✅

result.a.x = 5;
console.log(o1.a.x) // 5 🚫
```

Comment (nzakas) [reply lib/config/flat-config-schema.js:81]: Good point.

File: lib/config/flat-config-array.js [hunk 31]
```diff
@@ -0,0 +1,125 @@
...
+ *      name.
+ */
+function splitPluginIdentifier(identifier) {
+    const parts = identifier.split("/");
+
+    return {
+        objectName: parts.pop(),
+        pluginName: parts.join("/")
+    };
+}
```
[the remark this answers is not in the document]
Comment (mdjermanovic) [reply lib/config/flat-config-array.js:38]: The same logic is used in rule validator:

https://github.com/eslint/eslint/blob/86b7bddb421031db1a5c517a6ed6d1f23b388ba7/lib/config/rule-validator.js#L24-L35

so it would look for `es-syntax` rule in `node/no-unsupported-features` plugin, instead of `no-unsupported-features/es-syntax` rule in `node` plugin.

[the remark this answers is not in the document]
Comment (nzakas) [reply lib/config/flat-config-array.js:38]: So far this seems to be working okay with my tests. But again, this is just the first draft, so we have plenty of time to find and address errors. It will be easier to find once I start integrating.

**Review by mdjermanovic — APPROVED** [review]
LGTM!

I'm not sure if the feature of unignoring files that are ignored in the base config (or a shared config) is supported, but we can discuss that when we start integrating.
